aws_lambda_builders/workflows/nodejs_npm/utils.py

```diff
@@ -48,3 +66,189 @@ def abspath(self, path):
+        -------
+        dict
+            Dict of local dependencies key->value if any, empty otherwise
+        """
+        with osutils.open_file(manifest_path) as manifest_file:
+            manifest = json.loads(manifest_file.read())
+
+            if "dependencies" in manifest:
+                return dict(
+                    (k, v) for (k, v) in manifest["dependencies"].items() if DependencyUtils.is_local_dependency(v)
+                )
+
+            return {}
+
+    @staticmethod
+    def is_local_dependency(path):
+        """
+        Helper function to check if a dependency is a local package
+
+        Parameters
+        ----------
+        path : str
+            Path to check
+
+        Returns
+        -------
+        boolean
+            True if a local dependency, False otherwise
+        """
+        try:
+            return path.startswith("file:") or path.startswith(".")
+        except AttributeError:
+            return False
+
+    @staticmethod
+    def package_dependencies(manifest_path, scratch_dir, packed_manifests, osutils, subprocess_npm):
+        """
+        Recursively packages NPM dependencies, including local ones.
+        Handles circular dependencies and skips already processed ones.
+
+        Parameters
+        ----------
+        manifest_path : str
+            Path to the manifest to package
+        scratch_dir : str
+            Path to the scratch directory
+        packed_manifests : dict
+            Dict of hashes pointing to packed manifest tar used to avoid
+        osutils : OSUtils
+            OS utils
+        subprocess_npm : SubprocessNpm
+            NPM process
+
+        Returns
+        -------
+        string
+            Path to the packed tar file
+        """
+        manifest_path = osutils.normpath(manifest_path)
+        LOG.debug("NODEJS processing %s", manifest_path)
+        manifest_hash = str(abs(hash(manifest_path)))
+        if manifest_hash in packed_manifests:
+            # Already processed or circular dependency
+            # If empty, it will be removed from the manifest
+            return packed_manifests[manifest_hash]
+        packed_manifests[manifest_hash] = ""
+        manifest_dir = osutils.dirname(manifest_path)
+        manifest_scratch_dir = osutils.joinpath(scratch_dir, manifest_hash)
+        manifest_scratch_package_dir = osutils.joinpath(manifest_scratch_dir, "package")
+        osutils.makedirs(manifest_scratch_package_dir)
+
+        # Pack and copy module to scratch so that we don't update the customers files in place
+        DependencyUtils.ship_module(manifest_dir, scratch_dir, manifest_scratch_dir, osutils, subprocess_npm)
+
+        # Process local children dependencies
+        local_dependencies = DependencyUtils.get_local_dependencies(manifest_path, osutils)
+
+        for (dep_name, dep_path) in local_dependencies.items():
+            if dep_path.startswith("file:"):
+                dep_path = dep_path[5:].strip()
```

Comment:
nit: parsing path from a `package.json` `dependencies` value seems well-defined; could be refactored to its own method

Reply:
I'll move this to `get_local_dependencies` so that it returns them all in the same format.